### DISCUSSION TITLE
How to mark all fields in a type as optional.

### DISCUSSION THREAD

**Discussion by dhruv-nudge** [sinclairzx81/typebox#835] [Q&A] [answered]
In zod, there was option of .deepPartial() that basically makes all the fields as optional.
I am looking for a similar solution, which will allow me to mark all fields as optional, without explicitly wrapping every field with Type.Optional.

**Comment by sinclairzx81** [comment, marked as answer]
@dhruv-nudge Hi,

TypeBox doesn't provide a built in deepPartial method (as there is no equivalent TS utility type). In TypeScript however, the expectation is generally to leverage the type system programmatically produce an appropriate mapped type. TypeBox follows a similar reasoning, but is a bit more complex as you will need to map both runtime and static type.

The following programmatically implements a PartialDeep type.

[TypeScript Link Here](https://www.typescriptlang.org/play/?ssl=50&ssc=53&pln=49&pc=1#code/JYWwDg9gTgLgBAbzgFQJ5gKYHECuBDKAEwBoV0NTkBlAYwAsMQ9KBJAOxgygGcMaZKAVTbAIbSgHkARgCs+AlAAUCMYHgA2lRVAiZYwDN1IBRAG4b8nOAF84AMx0g4AcgAC3YGxrq8wKAHoYcikIAA9nACgI-384AFoExKTklNS09IzMrNTo2ORlfQ0AEQwMMG1dLlVDXPjs+obGpvSIjFDIWDggzCUVNXUSsoq9au4AHmQ4Ns42Qm4lHRGDbgA+OABeRAi4OABtAGk4TzgAawxUCDsUAF0ALl7CgdKwCYPrlYjrCLscL1UxOAFVTFZ7DKrLCZTUIzOYLSr6QwrAAUYEW4MM92QAEpMUD+oNymiEeNkGsENs4FAMDAcFA2HBpHJ+AA6ADm1IkAHc2GDYKgAHJ4ECGFFE0ZY5lUwg4GgYJFIvA0GikM6oLEbMkUnZUml0xBwZmGxXKvaqu6Avogsqi+GjXZm9VfHbWUgIazqvDzNgYUxcT61ZqBoPBnIxB7Ap5lABKhhgAZDCcTDVa7Wg8G6GHD+OeMe4MEh0wws3m1HojDwu2upAAgkqoTCS7QGExKxs9u820iKZNC8W9p47Fw4AAZetF2Gl5vMA2GgdDqNjvuT8uV64UgD8WatYFz+ajpF2huZtZN+UtkZew5W7wp9xPESx31+-FE9Lx293Beh48bZZb7yRKk83uQ9DWQa4cS3C9P1JLZtWpWl6SAmBmSYMAkW4P88A1C1HgJDCsKxD0vR9P0vjDJNKKo0M8nfC942oxjKJTDp03IKCCS-BsUCbcs1nWbtFwndhOB4eQxjnKA4CoNZN2QESuF4fgJjoglPxktZbkE3sJ2EV8JLYQcpJkuA5L0sQVPPNTYzGDS4C0nYe2-JdGXEyTpNkqCJlc5SzAsPBOEsvDQTFCENJWTTuxYtN7Gff43ys54uJ-HisORTCp0xSCz2CsoJk1eDdXpLsdkc8hcAIQhmRYbgFLE-gCKndU5PIaqOEU+QkVUnNY0a8tmQ0dQJDsIj7K1Uq0EwCqiGq7hzLYPqmGashMGZeausS6Neoy-q8DYVBhtGhzSpOybsHwGaap8mBFrwZazuZOikQe66NtywlbWWW7mVRT7DCI9VjpOk6dqYCliLgb1fSgf0KKY+GE1qQRuDwdkGIRjGmgiGgxDzOBqzbbrrRe2ROqQMMPDYVl1EzQgyhgOgKVCTFWv5HAQCkLgkSxYgKVQFnVrZjmuZ5ikAC8BYwZkhc5qBud591H2x3H4AAIUJzb0JJpkbvJ2IQBwdRVDAGm4DpsAGYpPBJeZV7yRO5mVqlmWRd5k7+ad6X2dl+Xxolz2Xbl0XnWDuApBtu3xsdh7A9992bdj0Odn9mPvddil3U+IigA)

```typescript
import { TypeGuard, Type, TSchema, TIntersect, TUnion, TObject, TPartial, TProperties, Evaluate } from '@sinclair/typebox'

// -------------------------------------------------------------------------------------
// TPartialDeepProperties
// -------------------------------------------------------------------------------------
export type TPartialDeepProperties<T extends TProperties> = {
  [K in keyof T]: TPartialDeep<T[K]>
}
function PartialDeepProperties<T extends TProperties>(properties: T): TPartialDeepProperties<T> {
  return Object.getOwnPropertyNames(properties).reduce((acc, key) => {
    return { ...acc, [key]: PartialDeep(properties[key]) }
  }, {}) as never
}
// -------------------------------------------------------------------------------------
// TPartialDeepRest
// -------------------------------------------------------------------------------------
export type TPartialDeepRest<T extends TSchema[], Acc extends TSchema[] = []> = (
  T extends [infer L extends TSchema, ...infer R extends TSchema[]]
  ? TPartialDeepRest<R, [...Acc, TPartialDeep<L>]>
  : Acc
)
function PartialDeepRest<T extends TSchema[]>(rest: [...T]): TPartialDeepRest<T> {
  return rest.map(schema => PartialDeep(schema)) as never
}
// -------------------------------------------------------------------------------------
// TPartialDeep
// -------------------------------------------------------------------------------------
export type TPartialDeep<T extends TSchema> =
  T extends TIntersect<infer S> ? TIntersect<TPartialDeepRest<S>> :
  T extends TUnion<infer S> ? TUnion<TPartialDeepRest<S>> :
  T extends TObject<infer S> ? TPartial<TObject<Evaluate<TPartialDeepProperties<S>>>> :
  T
export function PartialDeep<T extends TSchema>(schema: T): TPartialDeep<T> {
  return (
    TypeGuard.IsIntersect(schema) ? Type.Intersect(PartialDeepRest(schema.allOf)) :
      TypeGuard.IsUnion(schema) ? Type.Union(PartialDeepRest(schema.anyOf)) :
        TypeGuard.IsObject(schema) ? Type.Partial(Type.Object(PartialDeepProperties(schema.properties))) :
          schema
  ) as never
}
// -------------------------------------------------------------------------------------
// Usage
// -------------------------------------------------------------------------------------
const A = PartialDeep(Type.Object({ // single depth
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number(),
}))

const B = PartialDeep(Type.Object({ // multiple depth
  a: Type.Object({
    x: Type.Number(),
    y: Type.Number(),
    z: Type.Number(),
  }),
  b: Type.Object({
    x: Type.Number(),
    y: Type.Number(),
    z: Type.Number(),
  })
}))
```
TypeBox tries to keep runtime and static implementations somewhat symmetric (so, taking note of the similarities between the TS types and associated mapping functions).  You should be able to copy and paste the above type into your project (and modify as required) to get a deepPartial.

Hope this helps
S

**Reply by dhruv-nudge** [reply]
Hey thanks for such a quick reply, I will go through this.
Just wanted to ask one more thing, If I am using compiled version of `PartialDeep(T)`, then there won't be any performance penalty, right?

**Reply by sinclairzx81** [reply]
Heya, 

> Just wanted to ask one more thing, If I am using compiled version of PartialDeep(T), then there won't be any performance penalty, right?

The above type shouldn't impact compiler performance as the return type for `PartialDeep(T)` is a TSchema (the same as if you had written the type manually). In mapping scenarios like this, if you ensure the type remains some variant of TSchema (TObject, TArray, TNumber, TString, etc), then you will get all the performance benefits of the compiler :)

Hope this helps
S

**Reply by dhruv-nudge** [reply]
it definitely helps. thanks a lot.

**Reply by Dhruv-Garg79** [reply]
Hey I just tried the above method you provided, but I am not getting the same result as in the typescript snippet.

```
export const userSchema = PartialDeep(
	Type.Object({
		id: Type.String(),
		name: Type.String(),
		email: Type.String(),
		phone: Type.String(),
		props_int: Type.Record(Type.String(), Type.Integer()),
		props_bool: Type.Record(Type.String(), Type.Boolean()),
		props_text: Type.Record(Type.String(), Type.String()),
		props_float: Type.Record(Type.String(), Type.Number()),
		updated_at: Type.Date(),
		created_at: Type.Date(),
	}),
);

const compiledSchema = TypeCompiler.Compile(userSchema);
export type TUser = Static<typeof userSchema>;
```

<img width="469" alt="Screenshot 2024-04-20 at 1 05 59 PM" src="https://github.com/sinclairzx81/typebox/assets/40490622/0a7c8cf9-9c4a-42b1-9b09-44cd2e9f808b">

**Reply by darthsalad** [reply]
@sinclairzx81 In the example you provided above, objects like you mentioned are converted correctly,
```typescript
const TT = Type.Object({
	a: Type.Object({
		x: Type.Number(),
		y: Type.Number(),
		z: Type.Number(),
	}),
});
```

type: 
```typescript
const TTPartialDeep: TObject<{
    a: TOptional<TObject<{
        x: TOptional<TNumber>;
        y: TOptional<TNumber>;
        z: TOptional<TNumber>;
    }>>;
}>
```

but objects which don't have nested fields like these:
```typescript
const X = Type.Object({
	c: Type.Number(),
});
```

are not converted correctly and give this type inference:
```typescript
const XPartialDeep: TObject<{
    c: TOptional<TObject<{}>>;
}>
```

Can you explain why is this the case?

**Reply by sinclairzx81** [reply]
@dhruv-nudge @Dhruv-Garg79 @darthsalad Hiya,

It appears there was a bug in the DeepPartial implementation provided. The implementation was taken from the TypeBox `example/prototypes/partial-deep.ts` prototype (which hasn't been updated in a while). I've since updated the implementation snippet (above) to support single depth Partial (as noted by @darthsalad), have also checked in the prototype implementation at https://github.com/sinclairzx81/typebox/blob/master/example/prototypes/partial-deep.ts for future reference. You can test the implementation at the following URL

[TypeScript Link Here](https://www.typescriptlang.org/play/?ssl=50&ssc=53&pln=49&pc=1#code/JYWwDg9gTgLgBAbzgFQJ5gKYHECuBDKAEwBoV0NTkBlAYwAsMQ9KBJAOxgygGcMaZKAVTbAIbSgHkARgCs+AlAAUCMYHgA2lRVAiZYwDN1IBRAG4b8nOAF84AMx0g4AcgAC3YGxrq8wKAHoYcikIAA9nACgI-384AFoExKTklNS09IzMrNTo2ORlfQ0AEQwMMG1dLlVDXPjs+obGpvSIjFDIWDggzCUVNXUSsoq9au4AHmQ4Ns42Qm4lHRGDbgA+OABeRAi4OABtAGk4TzgAawxUCDsUAF0ALl7CgdKwCYPrlYjrCLscL1UxOAFVTFZ7DKrLCZTUIzOYLSr6QwrAAUYEW4MM92QAEpMUD+oNymiEeNkGsENs4FAMDAcFA2HBpHJ+AA6ADm1IkAHc2GDYKgAHJ4ECGFFE0ZY5lUwg4GgYJFIvA0GikM6oLEbMkUnZUml0xBwZmGxXKvaqu6Avogsqi+GjXZm9VfHbWUgIazqvDzNgYUxcT61ZqBoPBnIxB7Ap5lABKhhgAZDCcTDVa7Wg8G6GHD+OeMe4MEh0wws3m1HojDwu2upAAgkqoTCS7QGExKxs9u820iKZNC8W9p47Fw4AAZetF2Gl5vMA2GgdDqNjvuT8uV64UgD8WatYFz+ajpF2huZtZN+UtkZew5W7wp9xPESx31+-FE9Lx293Beh48bZZb7yRKk83uQ9DWQa4cS3C9P1JLZtWpWl6SAmBmSYMAkW4P88A1C1HgJDCsKxD0vR9P0vjDJNKKo0M8nfC942oxjKJTDp03IKCCS-BsUCbcs1nWbtFwndhOB4eQxjnKA4CoNZN2QESuF4fgJjoglPxktZbkE3sJ2EV8JLYQcpJkuA5L0sQVPPNTYzGDS4C0nYe2-JdGXEyTpNkqCJlc5SzAsPBOEsvDQTFCENJWTTuxYtN7Gff43ys54uJ-HisORTCp0xSCz2CsoJk1eDdXpLsdkc8hcAIQhmRYbgFLE-gCKndU5PIaqOEU+QkVUnNY0a8tmQ0dQJDsIj7K1Uq0EwCqiGq7hzLYPqmGashMGZeausS6Neoy-q8DYVBhtGhzSpOybsHwGaap8mBFrwZazuZOikQe66NtywlbWWW7mVRT7DCI9VjpOk6dqYCliLgb1fSgf0KKY+GE1qQRuDwdkGIRjGmgiGgxDzOBqzbbrrRe2ROqQMMPDYVl1EzQgyhgOgKVCTFWv5HAQCkLgkSxYgKVQFnVrZjmuZ5ikAC8BYwZkhc5qBud591H2x3H4AAIUJzb0JJpkbvJ2IQBwdRVDAGm4DpsAGYpPBJeZV7yRO5mVqlmWRd5k7+ad6X2dl+Xxolz2Xbl0XnWDuApBtu3xsdh7A9992bdj0Odn9mPvddil3U+IigA)

Let me know if this helps
S

**Reply by Dhruv-Garg79** [reply]
Hey @sinclairzx81 thanks a lot, it did help. Just one more issue is coming up for enums. The enums are becoming never
<img width="502" alt="Screenshot 2024-04-26 at 3 26 51 PM" src="https://github.com/sinclairzx81/typebox/assets/40490622/0027d342-811e-4036-98ef-7060d97be534">

The status field was passed to deepPartial using `type.enum`
